components/support/rust-logs/src/rust_logger.rs

```diff
@@ -0,0 +1,68 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+//! Rust Logger implementation
+//!
+//! This is responsible for taking logs from the rust log crate and forwarding them to a
+//! foreign_logger::Logger instance.
+
+use crate::foreign_logger::Logger as ForeignLogger;
+use parking_lot::RwLock;
+use std::sync::Once;
+
+// ForeignLogger to forward to
+static RUST_LOGGER: Logger = Logger::new();
+// Handles calling `log::set_logger`, which can only be called once.
+static INIT: Once = Once::new();
+
+struct Logger(RwLock<Option<Box<dyn ForeignLogger>>>);
+
+impl Logger {
+    const fn new() -> Self {
+        Self(RwLock::new(None))
+    }
+
+    fn set_foreign_logger(&self, foreign_logger: Option<Box<dyn ForeignLogger>>) {
+        *self.0.write() = foreign_logger;
+    }
+}
+
+impl log::Log for Logger {
+    fn enabled(&self, _: &log::Metadata<'_>) -> bool {
+        true
+    }
+
+    fn log(&self, record: &log::Record<'_>) {
+        if let Some(foreign_logger) = &*self.0.read() {
+            foreign_logger.log(record.into())
+        }
+    }
+
+    fn flush(&self) {}
+}
+
+pub fn set_foreign_logger(foreign_logger: Option<Box<dyn ForeignLogger>>) {
+    if let Some(foreign_logger) = &foreign_logger {
+        foreign_logger.log(crate::foreign_logger::Record {
+            level: crate::foreign_logger::Level::Warn,
+            target: "log-test".to_string(),
+            message: "in set_foreign_logger".to_string(),
+        });
+    };
+    INIT.call_once(|| {
+        log::set_logger(&RUST_LOGGER).expect(
+            "Failed to initialize rustlogs::Logger, other log implementation already initialized?",
+        );
+        log::set_max_level(log::LevelFilter::Debug);
```

Comment:
Hmm Do we want to manually set this there? If we do, is it possible that this races with the app calling `set_max_level`? I imagine we **always** want to keep the max level that the app sets, as opposed to the level set by the initializer. Can we add logic here to **not** set the max level if it's already been set by the app? We may be able to read the max level from https://docs.rs/log/latest/log/fn.max_level.html

But I'm not sure what it defaults too, and how we would differentiate between the app setting it, and it being the default value. Another option is to enforce that the app calls this API on the main thread, that's risky though since we use a `RwLock` 🤔 Thoughts?

Reply:
Good point.  The reason this is here is that Android doesn't set the max level and that results in way too many log entries being emitted.  I'll change it so we remember if the app has set a max level and only set a default if that hasn't happened.

I guess there's still a race if you had 1 thread calling `set_logger` and another calling `set_max_level()`, but that seems like something we can ignore.